src/Common/src/System/Security/Cryptography/Asn1V2.cs

```diff
@@ -0,0 +1,5246 @@
+            if (parsedLength > int.MaxValue)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            if (minimalRepresentation && parsedLength < MultiByteMarker)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            length = (int)parsedLength;
+            return true;
+        }
+
+        internal (Asn1Tag, int?) ReadTagAndLength(out int bytesRead)
+        {
+            if (TryPeekTag(_data.Span, out Asn1Tag tag, out int tagBytesRead) &&
+                TryReadLength(_data.Slice(tagBytesRead).Span, _ruleSet, out int? length, out int lengthBytesRead))
+            {
+                int allBytesRead = tagBytesRead + lengthBytesRead;
+
+                if (tag.IsConstructed)
+                {
+                    // T-REC-X.690-201508 sec 9.1 (CER: Length forms) says constructed is always indefinite.
+                    if (_ruleSet == AsnEncodingRules.CER && length != null)
+                    {
+                        throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                    }
+                }
+                else if (length == null)
+                {
+                    // T-REC-X.690-201508 sec 8.1.3.2 says primitive encodings must use a definite form.
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                }
+
+                bytesRead = allBytesRead;
+                return (tag, length);
+            }
+
+            throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+        }
+
+        private static void ValidateEndOfContents(Asn1Tag tag, int? length, int headerLength)
+        {
+            // T-REC-X.690-201508 sec 8.1.5 excludes the BER 8100 length form for 0.
+            if (tag.IsConstructed || length != 0 || headerLength != EndOfContentsEncodedLength)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+        }
+
+        private static ReadOnlyMemory<byte> SeekEndOfContents(
+            ReadOnlyMemory<byte> source,
+            AsnEncodingRules ruleSet,
+            int initialSliceOffset = 0)
+        {
+            ReadOnlyMemory<byte> cur = source.Slice(initialSliceOffset);
+            int totalLen = 0;
+
+            while (!cur.IsEmpty)
+            {
+                AsnReader reader = new AsnReader(cur, ruleSet);
+                (Asn1Tag tag, int? length) = reader.ReadTagAndLength(out int bytesRead);
+                ReadOnlyMemory<byte> nestedContents = reader.PeekContentBytes();
+
+                int localLen = bytesRead + nestedContents.Length;
+
+                if (tag == Asn1Tag.EndOfContents)
+                {
+                    ValidateEndOfContents(tag, length, bytesRead);
+
+                    return source.Slice(0, totalLen + initialSliceOffset);
```

Comment:
This seems to have skipped backwards past initialSliceOffset. Please verify/comment on what this is supposed to be doing.